### ISSUE TITLE
`variantModule` export is missing

### ISSUE THREAD

**Issue by ohana54** [paarthenon/variant#7]
Hi, thank you for this very promising library!

I'm trying to follow the example and use the handy `variantModule` function, but it seems like it's missing?
I'm using the following import statement:
`import { variantModule } from 'variant'`

The installed library version is 2.0.0.

Thanks :)

**Comment by paarthenon** [comment]
Ah damn! This is what I get for testing in local paths. Fixing now. In the meantime, you can do this:

```typescript
import {variantModule} from 'variant/lib/variant';
```

**Comment by paarthenon** [comment]
Thank you for doing some involuntary user testing for me 🥇 

This is fixed in variant `2.0.1` published nowish. 

And thank you for trying out the library. If you don't mind my asking, how'd you find it? I have been overhauling in prep for a 2.0 announcement but I haven't spammed reddit yet :).

**Comment by ohana54** [comment]
Thanks for the fix!

I was getting tired of the union boilerplate in my project and was looking on npm for something that would fit my use case. I just searched the words "typescript adt" :)

I have to be honest, after about an hour I switched to the `unionize` library, because I couldn't get `isVariantOf` to work as I wanted it.
For example, if we take the `Animal` variant module from the intro docs, I wanted the equivalent of:
```ts
const dogInstance = Animal.dog({ name: 'Rex' })
const isDog = Animal.is.dog(dogInstance)
```
There's a good chance it's possible and I just missed something :)

Also, I personally prefer the `default` key in a `match` function call, rather than `matchElse` that takes another callback.

This is a bit off-topic to this issue, but wanted to share my experience and feedback nonetheless.

**Comment by paarthenon** [comment]
Woah, I didn't know about unionize, it's funny (and surreal) to see how we made almost the same design decisions but along slightly different paths. If you have a minute to spare, I'd love to offer an explanation and pick your brain a little more on feedback.

> For example, if we take the Animal variant module from the intro docs, I wanted the equivalent of:
> ```typescript
> const dogInstance = Animal.dog({ name: 'Rex' })
> const isDog = Animal.is.dog(dogInstance)
> ```

You can do this with `dogInstance.type === 'dog'`. Variants are valid TypeScript discriminated unions, the same rules apply. I didn't tend to include options that have native ways of doing it that actually make sense to avoid being redundant, but it's great user feedback to hear that a "isDog" function is expected. I'll have to reconsider. Part of the reason was also not wanting to confuse users. I was also afraid if there's an "isDog" function most people will assume it's doing property validation which is _literally impossible right now._ `dogInstance.type === 'dog'` looks as fragile as it is :).

 - `isVariantOf`, by contrast, checks if an object is an instance of a given union (not case). Remember that this library can destructure and rebuild subsets of variants. isVariantOf is a user-defined type guard that can quickly to one of these subsets. If you have an `Action` type, you might do `isOfVariant(action, GameAction)` and then pass that action, now a `GameAction`, to your gameReducer.

It seems like the main difference between `variant` and `unionize` is in the user interface, and how obsessive we each are about your types and type inference. `variant` focuses on letting you build up these collections and also manipulate them as you wish to then pass to an agnostic `match` function (polymorphic variants), while unionize is much more focused on building up a single variant (direct ADTs). There are some consequences to this, which will come up in this next part

## `matchElse`

> Also, I personally prefer the default key in a match function call, rather than matchElse that takes another callback.

I'm 100% with you on this. Here's why I eventually went this way instead. Look at the 'else' argument. It's well-typed. If you do matchElse on `Animal` and you handle `dog` and `cat`, then the else function will be *guaranteed* to receive a `snake` and the type of the variable will reflect that. Even if we had the extended `Animal` object in play with `bird` and `pegasus` as valid options. The else block will understand that *it does not need to handle `dog`* The `default` property from unionize doesn't currently do this.
 - Variant's `match` will also let you return different types of elements and let their types flow. In unionize, if you want to match on a variant and one option returns a string you have to always return a string. That isn't the case here. This means you can match on something and not clobber the types of the content you are throwing around. 

The beauty here is you can write a match function to translate from variant A to variant B, then handle some of its cases of B in a match and scoop up the remaining two B cases in the default handler, where you can use another match on the remaining elements and the handler will be restricted to just the remaining cases of B, all without you ever adding a type annotation. I know this sounds silly but it actually comes up in the *Entity -> DTO -> ViewModel* flow. 

However, using `default` like unionize was the way I originally explored it and I do quite like the aesthetic, but I'm not sure if you can use the inferred type of an object to restrict the type of a property of that same object, the well-typed 'default' prop. Which means I would lose the fact that if I layer another `match` in my default case, it's smart about which cases I still need to handle. What I could do is merge `match` and `matchElse` with an overload so you can do this 

```typescript
const damage = match(animal, {
    cat: ({daysSinceDamage}) => daysSinceDamage,
}, rest => 0);
```
Woah yeah I like that. I also plan on ruthlessly stealing unionize's idea of using `{}` for empty bodied variants. I feel stupid for not seeing that as an option and using `nil` 🤣. Viva la open source. 

## Feedback
> This is a bit off-topic to this issue, but wanted to share my experience and feedback nonetheless.

Not at all, I'm very thankful to you for doing so. I did a similar comparison of libraries when someone brought up safety-match in issue #2 . Did the explanations make sense? 

### So in summary I'm really curious

 1. Does `animal.type === 'dog'` seem sufficient as a check? Would you like a more user-friendly named function?
 2. How does that proposed change to `match` strike you?
 3. How compelling are these explanations for your uses? Obviously these things matter to me, but they might not matter to you and that's ok!

This is in no way a popularity contest and you are absolutely welcome to use unionize if it's a better fit for your project and style, but I wanted to show what makes `variant` special compared to the others :).

**Comment by paarthenon** [comment]
Hi @ohana54 I ended up making these changes. They are available on the beta release, `npm i variant@beta`

I added the function `isType` which takes a possible variant and a type specifier -- a `string` or the variant constructor function directly. 
```typescript
test('IsType UDTG', () => {
    const kerb = Animal.dog({name: 'Kerberos', favoriteBall: 'yellow'});

    if (isType(kerb, 'dog')) {
        expect(kerb.favoriteBall).toBe('yellow');
    } else {
        fail('isType did not register kerb as a dog');
    }
});
```
You can use either:

```typescript
const isCat = isType(kitty, 'cat');
const isCat2 = isType(kitty, Animal.cat);
```

**Comment by ohana54** [comment]
Thank you for taking the time to respond :)

To answer your summary:

1. I prefer not using strings, so the `isCat2` example is more to my liking.
There is still an issue here I just thought of - as far as I understand the type guard won't work well with the array filter function:
```ts
const cats = animals.filter(animal => isType(animal, Animal.Cat))
```
Since I'm creating a new function, I need to "redeclare" the type guard for it, in order for `cats` to be an array of cats and not animals.
This does work with `unionize`, since the `isType` helper exists per variant:
```ts
const cats = animals.filter(Animal.is.Cat)
```

My personal preference is to have the `is` helper on the variant itself, so I won't have to import another function whenever I need it.
If you don't like this approach, maybe it can still work by making `isType` curried and accepting type first, data last.
WDYT?

2. About `match`:
	1. The overload is indeed a welcome addition!
	1. I understand the tradeoff of `default` vs. the better typed else callback, didn't know that. I'm used to the pattern matching in ReasonML where the default clause is not typed this way by design, so it never bothered me. It's a nice touch although specifically for me I'm not sure I'll need this better typing with my coding style.
	1. I didn't know about the limited return value type in `unionize`, that's a bummer! I don't have this use case at the moment, but I do fear it might be too limiting in the future.
	1. Due to all the above, I'm indeed leaning back to using `variant`!

I discovered an issue with potential typos in `match` that is not covered in `variant`. Consider the following case:
```ts
const name = match(
    animal,
    {
      Cat: () => 'cat',
      Dogg /* typo, should be Dog */: () => 'dog',
    },
    () => 'snake',
  )
```
This will not trigger any type errors, and will be a runtime bug.
Do you think it can be fixed?

3. The explanations are well put and I appreciate you taking the time to provide them, it was really helpful and I'm enjoying the discussion :)

**Comment by paarthenon** [comment]
>   There is still an issue here I just thought of - as far as I understand the type guard won't work well with the array filter function:
> 
> ```ts
> const cats = animals.filter(animal => isType(animal, Animal.Cat))
> ```

*goooood* point. I've added an overload to do what you describe

```ts
const cats = animals.filter(isType(Animal.cat));
const dogs = animals.filter(isType('dog'));
```

### Attaching the function to a variant

> My personal preference is to have the is helper on the variant itself, so I won't have to import another function whenever I need it. (`Animal.is.dog`)

It's a tempting path to follow that pattern, but it would require giving up part of this project's philosophy. I'm hesitant to add metadata to variant modules because I want users to be able to group and regroup the constructed modules on the fly without *needing* to go through my functions. I don't want to create a "blessed" space, I want to work with unions made by vanilla TS code or even other libraries. Hence the module really is a POJO. 

```ts
// I like being able to do this
const Action = {...GameAction, ...MenuAction, ...DebugAction};

// and some people really value being able to do this
const {dog, cat} = Animal;
const cerb = dog({name: 'Cerberus'});
```

**However**, there is certainly room for magic on each _case's constructor._ I could give you something like this:
```ts
const isDog = Animal.dog.match(x)
```
If that is a better fit than the current `isType` implementation. You know... now that I see this, I might do this anyway because it provides a hook for user-supplied validation functions that *can* then inspect object properties. This is just a mockup, but something resembling...

```ts
const Animal = variantModule({
    dog: fields<{name: string, favoriteBall?: string}>(),
    cat: fields<{name: string, daysSinceDamage: number}>({
        validator: c => c.daysSinceDamage >= 0,
    }),
    snake: ...
});
```
where `Animal.cat.match` would quietly run the validator predicate. Though I should *probably* not use the term match there again :). Perhaps `test`. Naming is hard. 

### matching `Dogg`

Yeah! You're absolutely right, but as far as I know this is is a TypeScript limitation. TypeScript cannot restrict something that extends a `Partial<T>` of some kind. Remember that every partial is also satisfied by the `{}` which is also satisfied by *any object*. That's essentially why I'm anal about retaining exhaustiveness checking by default. There are a couple of ways to mitigate this:

 1. use ES6 constant literals: 

    ```ts
    match(animal, {
        [Animal.dog.type]: ...,
        ...
    });
    ```
 2. leverage the 'or' clause's well-typed nature. So let's say, for example, I have animals that don't have a name. Ants, maybe. The others all have a property `name: string` So if I want to write a function called `getName`, I can do the special handling for ants, and then I can pass the rest through the default case. 

    ```ts
    const getName = (animal: Animal) => match(animal, {
        ant: _ => 'Auntie',
    }, rest => rest.name);
    ```
    If I were to add a new animal that did not have a name, the or clause would no longer be valid and that would raise a compiler error. 

Unfortunately to some extent this is what happens any time you don't handle exhaustive branching, as I'm sure you're familiar if you're a fellow ML nerd where `if` is literally an expression :). 

****

I'm also enjoying it, thank you very much! I'm being very genuine when I say this conversation has made variant a better library, and I'm glad to hear that it may be a good fit for you.

I'm publishing the update to `isType` as we speak. If you once again install `variant@beta` you should receive `2.0.2-beta.2` which includes the change. Documentation will come once the syntax is more finalized.

**Comment by ohana54** [comment]
Thanks for the awesome reply, again :)

1. About attaching the function to a variant - I indeed meant the helper functions are on the constructor itself. `unionize` also has POJO variants and I fully agree with this design decision.

2. Do you perhaps know how the "matching `Dogg`" thing does work correctly with `unionize`? I would understand if the internal implementation is completely different that it is unachievable here.

**Comment by paarthenon** [comment]
> 2. Do you perhaps know how the "matching Dogg" thing does work correctly with unionize? I would understand if the internal implementation is completely different that it is unachievable here.

I do! It's not unachievable, just finicky. I have a suggestion. But first, explaining what's happening: There's a specific situation where if you restrict a type w/ some partials and you are giving it an object literal that literal is evaluated with explicitly limited options. That's what unionize is doing. However, you cannot do this if you want to *use* the literal type of the object that's being passed in as an inferred generic parameter, as I do for the else handler *and* for the return type of match (another thing unionize foregoes)

**Here's what I'm thinking: That could be a very good fit for `partialMatch`**. As things are partialMatch *can't* use the return-type-union trick because types only flow one way and I really wanted partialMatch to restrict the keys if it's already so loose, so it's a perfect place to shove this kind of default functionality. That would make `match` the exhaustive handler with freaky ML-level types and `partialMatch` would be a lighter form that would be essentially equivalent to unionize's match statement, with one small caveat:

Unionize's match forces you to **either** give a partial + default property _or_ a fully specified object. I already have `match` so I think it would make sense to have partialMatch *always* allow missing parameters and have *opt-in* default handling functionality. That way you can still just... do nothing. In the case of a mismatch with no default, `partialMatch` would simply return undefined. 

I think this makes sense but I could probably be convinced to make `partialMatch` also require a default if the handler is incomplete. I mostly don't like that idea because then it just feels like the match-else overload but with a different syntax and more occlusive types.

****

p.s. 

> POJO variants

So this is what I mean. When you use unionize:
![image](https://user-images.githubusercontent.com/10353928/97729285-f3ad9780-1aa8-11eb-85eb-071a789bed96.png)

vs. variant
![image](https://user-images.githubusercontent.com/10353928/97729490-2ce60780-1aa9-11eb-85e3-3d84262e2bba.png)

You can straight up enumerate a variant module and it'll work fine.

**Comment by paarthenon** [comment]
Variant `2.0.2-beta.3` is the new `variant@beta` version. This patch includes the update to `partialMatch` I described, if you'd like to check it out. This includes the strict literal property checking and a default property. 

![image](https://user-images.githubusercontent.com/10353928/97771938-5db15580-1b18-11eb-9bfb-e410a491edd4.png)

This patch also changes `variantModule` to use `{}` for empty-bodied variants instead of `undefined`.

**Comment by ohana54** [comment]
About the POJO support, I referred to the variant instances (`dog`, `cat`), not the "module" itself (`Animal`). I don't really care that the module itself is POJO although I understand where you're coming from.
In fact, if I can just add more keys, can I achieve the same API as `unionize`? Meaning somehow automatically adding `is.dog(instance)` and `is.cat(instance)` when having `dog` and `cat` constructors.

I'll go and check out the new `partialMatch`, will report back :)

**Comment by ohana54** [comment]
For some reason the return type of `partialMatch` is `any`:
![image](https://user-images.githubusercontent.com/3693735/97779339-4b193980-1b75-11eb-81f8-07cc3e81a56a.png)


BTW, I still think from a user's perspective, the fact that `match` with the else callback supports arbitrary keys is a pitfall and can cause wrong runtime behavior.

**Comment by paarthenon** [comment]
Hey @ohana54 sorry for not responding I needed to catch up on some paid work and I also wanted to make sure I did your question justice so I spent some time trying to fix it again before diving in.

> For some reason the return type of partialMatch is any:

This is the problem. As far as I know, and I would welcome your help on this, it is impossible to simultaneously
 1. Restrict the keys of a partial.
 2. Use the type of the object literal satisfying that partial.

Consider such a function. Let's say you have 

```
export function match<T extends {type: string}, H extends Handler<...>>(object: T, handler: H): ReturnType<H[keyof H]> {
    ...
}}}
```
This is roughly what `match` looks like. Now we want to make it partial. The most direct approach would be to alter the type of the generic to extend the partial. 

 > H extends **Partial**<Handler<...>>

But that unfortunately doesn't help us. If we try out this code, the `handler` object doesn't have *any* autocomplete. That's because of the

 > TypeScript cannot restrict something that extends a `Partial<T>` of some kind. Remember that every partial is also satisfied by the `{}` which is also satisfied by any object.

comment from earlier. I should have said for generics, since that was what it was referencing. Anyway `{hello: 4}` extends `{}` which extends `Partial<T>` no matter what `T` is, and so we can't use the properties of `T` to restrict the object literal. That's autocomplete and exhaustiveness checking dead.

We could instead put the partial at the handler argument: 

> (object: T, hander: Partial<H>) {

We will now get keys! And they will be restricted to valid properties of the handler object (this can also be used to implement `default` behavior). BUT, we are now supplying an object literal that meets a contract, not supplying the underlying type for a generic to infer. By doing this, we are clobbering the literal's type and as far as TypeScript is concerned `handler` is Partial<H> and anything we give is just an implementation detail. This means I cannot know your return types, hence the `any` that you see (partialMatch currently works this way).

### _Pick your poison_

So that's why `matchElse` favors a well-typed return over the restricted keys. It's doing more work for you this way. Creating an else block that takes into account cases that have already been handled and can continue to do so as code is refactored is hard. Using property keys if you have to `[Animal.dog.type]:` to avoid mistyping is less hard. 

`match` currently supports both the return type and the restricted keys. So if I'm going to add an overload to support matchElse-like functionality it would ideally keep that behavior. This seems impossible, so, *which of the two* is least surprising and/or damning to lose. If you lose autocomplete and restricted keys, you may have occasional user fuckups and annoyance. If you lose the well-typed return value and start pushing `any` through user systems, you may truly break things.

That's essentially how I chose, but as always continued user feedback is a part of the pie. If that is still unsatisfactory I'd like to hear it. And if you should happen to have a fix that works for your use cases and the library functionality, I'd be all ears. 

## Approaches

I haven't given up on trying. There are two things I think have potential. One is to use the second parameter of the `Handler<T, U>` type. If you actually assign this you run into the same problem as unionize -- you can only return one type of data. If I can interpret this to be the union of possibilities there may be some potential, but I don't have high hopes on this.

The other thought is to try and write a generic higher order function that will narrow to the contextual requirement without user input. Ideally something that would look like this:

`match(object, partial({...})`

Though that's hard too.

**Comment by paarthenon** [comment]
> ## **Edit** I may be really dumb and just be wrong about this.

Okay so yes and no, but mostly yes. I was right about my explanation, but wrong in my thinking that was the only option. When a default type is made *mandatory* and a part of the object contract unless the full spec is provided this `Partial` issue disappears.

The new version of `variant@beta`, `2.0.2-beta.4` includes these changes. So as review:

## Forms of match: 

1.  Normal, exhaustive match:
    ```typescript
    const rating =(a: Animal) => match(a, {
        cat: _=> 1,
        dog: _=> 2,
        snake: _ => 3,
    });
    ```
 1. Partial match with default
    ```typescript
    const rating = (a: Animal) => match(a, {
        cat: _ => 1,
        default: _ => 2,
    });
    ```
 1. Partial match with 'else' handler
    ```typescript
    const rating = (a: Animal) => match(a, {
        cat: _ => 1,
    }, _ => 2);

For option 2, the parameter to the default case is the full union. 

For option 3, the parameter to the else case is the remaining cases unhandled by the partial

All of them *should* have functional return types.

I hope that works for you.

**Comment by ohana54** [comment]
Thanks, I checked the latest beta and the return type issue is indeed fixed :)

I got a bit lost in your last comments - the only way to work around the typo pitfall is computed keys?

Also, what do you think about my previous comment? Is there a good way I can add functions to the variant module? As I don't need it to be a POJO (I hope I'm not wrong about this approach)

**Comment by paarthenon** [comment]
> I got a bit lost in your last comments

Sorry, you can ignore everything that I put into the quote. It's wrong. I started writing this library at TS version 2.X and so when I was experimenting I learned some lessons that aren't true anymore. I just left it up in case it was helpful to me or someone in the future to see where my mind was as I was responding. TypeScript's mapped types and conditional types are way more powerful now than they were then.

I might be able to do the typo error. I'll mess with it later today. 

> Also, what do you think about my previous comment? Is there a good way I can add functions to the variant module? As I don't need it to be a POJO (I hope I'm not wrong about this approach)

It's complicated. It sort of works. I just spent a few minutes trying it.

Setup
```typescript

export type IsHandler<T extends VariantModule> = {
    [P in keyof T]: (obj: {}) => obj is VariantOf<T, P>;
}

export function isHandler<T extends VariantModule>(variantDef: T) {
    return Object.keys(variantDef).reduce((acc, key) => {
        return {
            ...acc,
            [key]: isType(key),
        }
    }, {}) as IsHandler<T>;
}

export function enhanced<T extends VariantModule>(variantDef: T) {
    return {
        ...variantDef,
        is: isHandler(variantDef),
    }
}

```

Execution
```
const _Animal = variantList([
    variant('dog', fields<{name: string, favoriteBall?: string}>()),
    variant('cat', fields<{name: string, daysSinceDamage: number}>()),
    variant('snake', (name: string, patternName?: string) => ({
        name,
        pattern: patternName ?? 'striped',
    })),
])
export const Animal = enhanced(_Animal);
export type Animal<T extends TypeNames<typeof _Animal> = undefined> = VariantOf<typeof _Animal, T>;
```

I had to set up an `_Animal` intermediary because if the `is` property is present then the `VariantOf` and `TypeNames` helpers don't know how to handle it. There are some other functions that won't work. However, `match` seems to handle it just fine.

**Comment by paarthenon** [comment]
> I might be able to do the typo error. I'll mess with it later today.

Hi I *believe* variant `2.0.2-beta.5` addresses this.

**Comment by ohana54** [comment]
Nice!!!
What I'm really missing in `unionize` is the ability to create a custom constructor with calculated properties (similar to the `snake` example above), I can't wait to replace it with `variant`!

The major thing that will change is the need to import `isType` and `match` everywhere - the solution you provided is nice, although I can't ask developers to be aware of the intermediary type, as it's too low-level when all they want to do is create a union.
I'll discuss this new API with the team and make sure they are all okay with it.

**Comment by paarthenon** [comment]
That's good to hear. As always they are welcome to post here if there are questions, suggestions, or requests.

I hadn't really considered the _imports_ as part of the ergonomics of the library, though perhaps I should have. What is it like for your setup? I'm using vscode with a project-level `tsconfig.json` and when I have variant installed as a library I can just start typing `match` and when I press enter to autocomplete the term it will also automatically include the import. That works for `isType` too. So with that flow it doesn't really cost me anything to have multiple imports, but of course everyone's setup and preferences are different.

Also, assuming that I do attach a UDTG like `isType` to the constructor function do you have any suggestions for naming? As things are I'm tempted to go `isType` so it'll be `if (Animal.dog.isType(x))` or `animals.filter(Animal.dog.isType)`. I admit it isn't as clean as `Animal.is.dog` but hey it's a start.

**Comment by ohana54** [comment]
Our setup is similar, vscode's auto import indeed helps.
We work in a monorepo, so the downside I can think of is having to add `variant` as a dependency in a specific package that needs to use `isType` or `match` on a variant that came from another package. If those functions were attached, there was no need to add the dependency.

As for naming, your suggestion is good :)
It does mean that `isType` and `match` won't be on the same object level, as `match` will be on `Animal`. I don't know if it really matters.

**Comment by paarthenon** [comment]
Ah, I see what you mean. If you don't mind my asking, roughly how many projects do you maintain in that monorepo? I don't know how you'll feel about this suggestion, but if you'd like to be able to import the match without manually importing from `variant` you could always repackage the functions you care about as part of your core libs. I haven't tried following this pattern since I just toss variant in most projects but everything is a pure function built on top of TS's unions so it *should* work.

```typescript
import {Animal, match} from '@appname/core';
```

> It does mean that isType and match won't be on the same object level, as match will be on Animal

I wasn't planning on putting `match` on the variant module (`Animal`), just `isType` on the constructors, since those are already magic.

**Comment by ohana54** [comment]
It's a single product divided into multiple packages. We do have some core libs, and I did actually put `unionize` in one of the core libs and re-exported it from there. Although I'm still not sure about it, I usually don't like to abstract 3rd party libs if I don't have a really good reason.

I don't think it's such a big deal whether the functions are separate or not. I'll get some more feedback from the team. I'm sure they will understand tradeoffs.

**Comment by ohana54** [comment]
Hey, it's been a few weeks but I'd like to update that we're going to replace unionize with variant :)
Are you planning to release a stable version soon, or do you want us to run with the beta for some time?

**Comment by paarthenon** [comment]
Hi @ohana54 thanks for following up. I'm pleased to hear that.

I'm sorry, in the time since I've had two losses in my family. I just haven't been on github. I'll be able to make a stable release soon.

**Comment by ohana54** [comment]
I'm so sorry to hear, please accept my deepest condolences.
Please put the version aside for now, it is not important, and let me know if I can be of any help.

**Comment by paarthenon** [comment]
I want to thank you for your patience. I've returned, and I've since made a stable release `variant@2.0.2` is now available with these changes.

**Comment by ohana54** [comment]
Thank you for everything, this thread was the best open-source interaction I've ever had.
We will do the migration in the coming weeks and return if we have any issues.

**Comment by paarthenon** [comment]
I can safely say the same :). Thank you for being so willing to discuss and work with me on these changes! It's been truly engaging experience. As the project grows I'd like to offer some visible acknowledgments, either in the form of a specific page or as a note of thanks in the changelog for the specific features that came from specific people. Would you or your organization like to be listed?

****

There are also a couple of new features I'm proud of that may help your team in the transition. Everything below this line is new: https://paarthenon.github.io/variant/docs/use/matching#match-helpers. The match helpers are one thing, but the builder-pattern `matcher` finally allows to handle multiple cases in one block, just like in a `switch` statement.

```ts
const describeAnimal = (animal: Animal) => matcher(animal)
    .when({
        snake: s => `${s.name} is enjoying the heat of the lamp on his ${s.pattern} skin`,
    })
    .when(['cat', 'dog'], ({name}) => `${name} is hoping you will pet their fur.`)
    .complete();
```

I know the TS-community has been looking for a good multi-match solution for some time, I'm very excited to hopefully be delivering one.

**Comment by ohana54** [comment]
Hey, not sure if there's an issue with mentioning my company name and prefer not to get into that, but feel free to mention me if you'd like :)

The multiple cases matcher looks very interesting, yet another thing I'm missing from a real language :)
I'm sure it'll come handy, I actually think I already needed it in parts of the code.

I went over the helpers, also nice additions!
One really tiny note, the term `just` is sometimes used for the `Maybe` ADT (Just/Nothing), there's a small chance it might be confusing. Lodash [uses the term `constant`](https://lodash.com/docs/#constant) for a similar functionality.

**Comment by paarthenon** [comment]
> Lodash [uses the term `constant`](https://lodash.com/docs/#constant) for a similar functionality.

So this is great. I apparently already implemented this functionality, *called it `constant`*, and have had it as a published export since a week before this issue was originally opened (Oct 20th).

So I guess... feel free to use `constant()`, lmao. Both are available.

**Comment by paarthenon** [comment]
Hi @ohana54 I was lamenting github's lack of ability to reach out directly and then I remembered this thread was still open. So I've got something undocumented I'm playing with that  is relevant to this conversation - the concept of a "remote", a wrapping type around the module. It consolidates the constructors into a sub-property, so now I no longer need to be concerned about the variant functions name-conflicting with the desired variants of a user. This can safely replicate the `is` syntax. 

```ts
const $Animal = remote(Animal);
test('remote', () => {
    expect($Animal.is.cat(cerberus)).toBeFalsy();
    expect($Animal.is.dog(cerberus)).toBeTruthy();
});

const dog = $Animal.new.dog({name: 'Fido'});
```
[See tests](https://github.com/paarthenon/variant/blob/master/src/variant.remote.spec.ts)

This is a stepping stone to the more advanced wrapper, `sequence`, which adds an ordering and extra functions like `.compare(element/type, element/type)`, `.get(index)` and `.index(element/type)`.

Sequence features two constructors, one takes in an existing module and a sorted list of types. The other works just like `variantList`.

[See tests (sequence)](https://github.com/paarthenon/variant/blob/master/src/variant.order.spec.ts).

It's still rather beta, but I thought you might have an interest in seeing that it's coming.

**Comment by ohana54** [comment]
Thanks for the update! I have to say we really got used to the global `match` and `isType` and haven't looked back since then :)

**Comment by paarthenon** [comment]
works for me :). Is there anything remaining in this thread that you wanted to talk about or am I free to close?

**Comment by ohana54** [comment]
You can definitely close, we're happy users ever since :)
